File: train/drive.py
#!/bin/env python3.2
# coding=utf-8


import math
import train.route


class DriveMark:
	'''
	Information on the driving
	'''

	def __init__ (self, time, distance, speed):

		self.time = time					# [s]
		self.distance = distance			# [m]
		self.speed = speed					# [km/h]


class DriveUnit():
	'''
	Elementary physics needed to ride the train
	'''

	def __init__ (self):

		self.maxspeed = 140		# [km/h]
		self.power = 2000		# [kW]
		self.edb = 1700			# [kW]
		self.weight = 155.5		# [t] 
		self.lenght = 79.2		# [m]
		self.wcount = 3			# number of wagons

		# ToDo: To determine the parameters directly from the specified train
		# (need database)

		#	vehicle		weight [t]	d		a		b		c
		#	----------------------------------------------------
		#	CD 471		62.7		0,15	
		#	CD 071		45.5		0,06	0,00033	0,0008	1,35	
		#	CD 971		47.3		0,06	0,00033	0,0008	1,35

		self.rho_a = 0.030624
		self.rho_b = 0.07424
		self.rho_c = 1.35
		self.rho_d = 14.973
		
		
#-------------------------------------------------------------------------------
#	F O R C E S

	def adhesion (self, velocity):
		'''
		Estimate of adhesion 
		http://homen.vsb.cz/~s1i95/mvd/mvd_tah.htm
	
		IN: velocity [km/h]
		OUT: adhesion coefficient [-]
		'''
		# Kother
		# return (9000.0/(math.fabs(velocity)+42.0) + 116.0)*0.001

		# Curtius-Kniffler
		return (7500.0/(math.fabs(velocity)+44.0) + 161.0)*0.001

		
	def dragForce (self, velocity):
		'''
		Drag of vehicles during moving
		
		IN: velocity [km/h]
		OUT: drag force [N]
		'''

		return (math.pow(velocity,2.0)*self.rho_a + velocity*self.rho_b + self.rho_c)*9.81


	def slopeDragForce (self, route, position):
		'''
		Drag force arising from the uphill / downhill
		
		IN: route, position [m]
		OUT: slope drag force [N]
		route ... description of the rail route
		'''
		#ToDo: implement resistive forces caused by cornering or a tunnel

		return self.weight * 9.81 * route.getSlope(position)
	
		
	def engineForce (self, velocity):
		'''
		Engine force
		
		IN: velocity [km/h]
		OUT: engine force [N]
		'''
		
		# Have you ever tried to divide by zero? (Well the case of velocity = 0
		# would be a different formula, but at the moment it is unknown.)
		U = velocity
		if (velocity < 20.0): U = 20.0

		return (3600 * self.power * self.adhesion(velocity))/U


	def uniformEngineForce (self, velocity):
		'''
		Throttle value for steady movement (acceleration = 0 m/s^2)
		
		IN: velocity [km/h]
		OUT: throttle_coefficient [-]
		'''
		
		# ToDo: So far only calculated for flat terrain! (slope = 0 ppt)

		return self.dragForce(velocity)/self.engineForce(velocity)


	def brakeForce (self, velocity):
	
		#aa = self.adhesion(velocity) * 9810 * self.weight
		bb = (self.adhesion(velocity)/0.15) * self.wcount * 20000 * 0.75
		return bb


#-------------------------------------------------------------------------------
#	M O T I O N

	def eqMotion (self, velocity, mode, route, position):
		'''
		Differential equation of motion

		IN: velocity [km/h], mode [-], route, position [m]
		OUT: a = dv/dt [m/s^2]
		Note: mode ... accelerator / brake lever
			route ... description of the rail route
			position ... it is needed to determine the climb / descent
		'''
		
		engine = lambda mode, velocity: self.engineForce(velocity) if mode == 1 else 0.0
		brake = lambda mode, velocity: self.brakeForce(velocity) if mode == -1 else 0.0
		drag  = self.dragForce(velocity) + self.slopeDragForce(route, position)
		
		return (engine(mode, velocity) - brake(mode, velocity) - drag)/(1000*(2.0*self.weight+self.rho_d))


	def deltaVelocity (self, dt, vA, mode, route, position):
		'''
		4th-order Runge–Kutta method
		http://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods
		
		IN:	dt ... time step [s]
			vA ... actual velocity [km/h]
			mode ... accelerator / brake lever
			route ... description of the rail route
			position ... it is needed to determine the climb / descent
		OUT: delta v [m/s]
		'''
		
		rkA = self.eqMotion (vA, mode, route, position)
		rkB = self.eqMotion (vA + (0.5 * rkA), mode, route, position)
		rkC = self.eqMotion (vA + (0.5 * rkB), mode, route, position)
		rkD = self.eqMotion (vA + rkC, mode, route, position)

		return (rkA + 2.0*rkB + 2.0*rkC + rkD)*dt/6.0


	def motion (self, route, position, distance, speedA, speedB):
		'''
		Train passing the section of route
		
		IN: route ... description of the rail route
			position ... it is needed to determine the climb / descent [m]
			distance ... length of section [m]
			speedA ... actual initial velocity [km/h]
			speedB ... new speed limit at the end of section [km/h]
		OUT: list of DriveMark (time = 0, distance, speed)		
		'''

		dt = 0.25
		xA = 0.0
		result = []
		
		# uniform movement
		if (speedA == speedB):
			result.append( DriveMark(0, position + distance, speedA) )
		
		# accelerated motion
		if(speedA < speedB):
			vA = speedA
			while (xA < distance) and (vA < speedB):
				vB = vA + 3.6*self.deltaVelocity(dt, vA, 1, route, position + xA)	# [km/h]
				xB = (dt*vB)/3.6 + xA												# [m]
				vA = vB
				xA = xB	
				result.append( DriveMark(0, position + xB, vB) )
		
		# deccelerated motion
		if (speedA > speedB):		
			vA = speedB
			result.append( DriveMark(0.0, position + distance, speedB) )
			while (xA < distance) and (vA < speedA):
				vB = vA + 3.6*self.deltaVelocity(-dt, vA, -1, route, position + distance - xA)	# [km/h]
				xB = (dt*vB)/3.6 + xA															# [m]
				vA = vB
				xA = xB	
				result.insert(0, DriveMark(0, position + distance - xA, vB) )
		
		return result

#-------------------------------------------------------------------------------		
# T R A I N   D R I V E R


	def deltaSpeedDistance (self, speedA, speedB, route, position):
		'''
		Returns the distance needed to change the speed (speedA --> speedB)
		
		IN: speedA ... initial speed [km/h]
			speedB ... final speed [km/h]
			route ... description of the rail route
			position ... it is needed to determine the climb / descent [m]
		OUT: distance [m]
		'''
		
		dt = 0.25
		vA = speedA
		xA = 0.0
		xB = 0.0

		if (speedA > speedB):
			mode = -1
		elif (speedA < speedB):
			mode = 1
		else:
			return 0	

		# The lambda tests whether already achieved the required speed
		condition = lambda mode, current, final: True if (mode * (current - final)) < 0.0 else False
		while (condition(mode, vA, speedB)):
			vB = vA + 3.6*self.deltaVelocity(dt, vA, mode, route, position)	# [km/h]
			xB = (dt*vB)/3.6 + xA											# [m]
			vA = vB
			xA = xB

		return xB
					

	def getOutlook (self, speedA, speedB, distance, route, position):
		'''
		Returns the speed, distance in order to optimize the passage section
		(type: speedA ​​> speedB, initial_speed < speedA)

		IN: speedA ... initial speed [km/h]
		speedB ... speed limit of next section [km/h]
		distance ... lenght of current section [m]
		route ... description of the rail route
		position ... it is needed to determine the climb / descent [m]
		OUT: speed [km/h], distance [m]
				
		Note: We would like to train first accelerated and then started to slow
		down the speedB.
		'''
		speedlimit = route.getSpeedLimit(position)
		
		# If a sufficient distance to slow is less than the length of the section.
		if (self.deltaSpeedDistance(speedA, speedB, route, position) < distance):
			i = 0
			suff_distance = 0.0
			speedC = speedA + i			
			while ((suff_distance < distance) and (speedC <= speedlimit)):
				i += 0.2
				suff_distance = 0.0
				speedC = speedA + i
				result_distance = self.deltaSpeedDistance(speedA, speedC, route, position)
				suff_distance += self.deltaSpeedDistance(speedC, speedB, route, position)
				suff_distance += result_distance
				
			return (speedC-0.2),  result_distance

		return speedA, 0.0


	def run (self, route):
		'''
		Passing route
		IN: route ... description of the rail route
		OUT: list of informations on the driving (time, distance, velocity
		'''	
		
		position = 0.0
		speed = 0.0
		passage = []

		# for each section of rail route
		for i in range(len(route.route)-1):
			distance = route.route[i+1].distance - position
			speedA = route.route[i].speedlimit
			speedB = route.route[i+1].speedlimit

			# uniform drive or acceleration on current section
			if (speedA <= speedB):
				passage += self.motion(route, position, distance, speed, speedA)
				position += distance
				speed = passage[-1].speed

			# We need to slow down, but we will try to pass as quickly as possible.
			if (speedA > speedB):
				# So we simply divide the section into two parts, we will try to
				# reach a maximum speed so that we can still safely slow down.
				speedC, tmpdistance = self.getOutlook(speed, speedB, distance, route, position)

				passage += self.motion(route,position,tmpdistance,speed,speedC)
				position += tmpdistance
				speed = passage[-1].speed

				passage += self.motion(route,position,distance-tmpdistance,speed,speedB)				
				position += distance - tmpdistance
				speed = passage[-1].speed

		# calculation of travel time
		for i in range(len(passage)-1):
			delta_v = (passage[i+1].speed - passage[i].speed) / 2.0 + passage[i].speed
			delta_distance = passage[i+1].distance - passage[i].distance
			passage[i+1].time = math.fabs(2 * delta_distance / delta_v) + passage[i].time

		return passage



if __name__ == "__main__":

	import doctest
	doctest.testmod()

File: cpp/src/drive.cc
#include "drive.hh"
#include <cmath>
#include <vector>
#include <tuple>
#include <cstdio>


using namespace std;


Drive::Drive(RailRoute &r): route(r) {

	maxspeed = 140;
	power = 2000;
	edb = 1700;
	weight = 155.5;
	lenght = 79.2;
	wcount = 3;
	rho_a = 0.030624;
	rho_b = 0.07424;
	rho_c = 1.35;
	rho_d = 14.973;
	
}

double Drive::engineForce (double velocity) {

	double U = velocity;
	if (velocity < 20.0) U = 20.0;

	for (int i=5; i > 0; i--) {
		if ((velocity > 20.0) and (velocity > maxspeed/i))
			U = maxspeed/i;
	}

	return (3600 * power * adhesion(velocity))/U;
}

double Drive::eqMotion (double velocity, double mode, double position) {
		
	double engine, brake, resist;
	
	if (mode == 1) {
		engine = engineForce(velocity);
		brake = 0.0;
	}
	if (mode == -1) {
		engine = 0.0;
		brake = brakeForce(velocity);	
	}	
	resist  = resistForce(velocity)+ resistSlopeForce(position);
	
	return (engine - brake - resist) / (1000 * (2.0 * weight + rho_d));
}

double Drive::deltaVelocity (double vA, double dt, double mode, double position) {
		
	double rkA, rkB, rkC, rkD;
			
	rkA = eqMotion (vA, mode, position);
	rkB = eqMotion (vA + (0.5 * rkA), mode, position);
	rkC = eqMotion (vA + (0.5 * rkB), mode, position);
	rkD = eqMotion (vA + rkC, mode, position);

	return (rkA + 2.0*rkB + 2.0*rkC + rkD)*dt/6.0;
}

void Drive::motion (double speedA, double speedB, double distance, double position) {

	double dt = 0.25;
	double xA = 0.0;

	if (speedA == speedB)
		drive.push_back ( DriveMark(0, position + distance, speedA) );
	
	if(speedA < speedB) {
		double vA = speedA;
		double vB;
		double xB;
		while ((xA < distance) and (vA < speedB)) {
			vB = vA + 3.6 * deltaVelocity(vA, dt, 1, position + xA);//1, route, )	# [km/h]
			xB = (dt*vB)/3.6 + xA;												//# [m]
			vA = vB;
			xA = xB	;
			drive.push_back( DriveMark(0, position + xB, vB) );
		}
	}
	
	if (speedA > speedB) {	
		double vA = speedB;
		double vB;
		double xB;
		vector<DriveMark>::iterator curr_pos = drive.end();
		drive.push_back( DriveMark(0.0, position + distance, speedB) );
		
		while ((xA < distance) and (vA < speedA)) {
			vB = vA + 3.6 * deltaVelocity(vA, -dt, -1, position + distance - xA);//, vA, -1, route, )	# [km/h]
			xB = (dt*vB)/3.6 + xA;															//# [m]
			vA = vB;
			xA = xB	;
			
			curr_pos = drive.insert (curr_pos, DriveMark(0, position + distance - xA, vB));
		}
	}
}

double Drive::deltaSpeedDistance (double speedA, double speedB, double position) {
		
	double dt = 1;
	double vA = speedA;
	double xA = 0.0; 
	double xB = 0.0;
	double vB;
	double mode;

	if (speedA > speedB) {
		mode = -1;
	}
	else if (speedA < speedB) {
		mode = 1;
	}
	else {
		return 0;	
	}

	while ( (mode * (vA - speedB)) < 0.0 ) {
		vB = vA + 3.6 * deltaVelocity(vA, dt, mode, position);//, vA, mode, route)	# [km/h]
		xB = (dt*vB)/3.6 + xA;//											# [m]
		vA = vB;
		xA = xB;
	}

	return xB;
}

tuple <double, double> Drive::getOutlookA (double speedA, double speedB, double distance, double position){

	double result_distance;
	double pos = position;
	double speedlimit = route.getSpeedLimit(pos);

	if (deltaSpeedDistance(speedA, speedB, pos) < distance) {
		double i = 0;
		double suff_distance = 0.0;
		double speedC = speedA + i;		


		while ((suff_distance < distance) and (speedC <= speedlimit)) {
			i += 0.2;
			suff_distance = 0.0;
			speedC = speedA + i;
			result_distance = deltaSpeedDistance(speedA, speedC, pos);
			suff_distance += deltaSpeedDistance(speedC, speedB, pos);
			suff_distance += result_distance;
		}
		
		return tuple <double, double> {(speedC-0.2),  result_distance};
	}

	return tuple <double, double> {speedA, 0.0};	
}

double Drive::funct (double CC, double speedA, double speedB, double distance, double position) {

	double currentA = deltaSpeedDistance(speedA, CC, position);
	double currentB = deltaSpeedDistance(CC, speedB, position);

	return distance - currentA - currentB;
}

tuple <double, double> Drive::getOutlookB (double speedA, double speedB, double distance, double position) {

	double C, fC;
	double pos = distance/2.0 + position;
	double speedlimit = route.getSpeedLimit(pos);

	if (deltaSpeedDistance(speedA, speedB, pos) < distance) {

		double A = speedA;
		double B = speedlimit;
		double fA = funct(A, speedA, speedB, distance, pos);
		double fB = funct(B, speedA, speedB, distance, pos);

		if (fA * fB < 0) {
			for (int i=0; i < 200; i++) {

				C = (A + B)/2;
				fC = funct(C, speedA, speedB, distance, pos);
				fA = funct(A, speedA, speedB, distance, pos);
				fB = funct(B, speedA, speedB, distance, pos);

				if (fA * fC > 0.0) {
					A = C;
				}
				if (fB * fC > 0.0) {
					B = C;
				}
				if (fC == 0.0) {
					return tuple <double, double> {C, deltaSpeedDistance(speedA, C, pos)};
				}
				if (fabs(A - B) < 0.2) {
					break;
				}
			}
			return tuple <double, double> {(C),  deltaSpeedDistance(speedA, C, pos)};
		} else {
			return tuple <double, double> {speedA, distance - deltaSpeedDistance(speedA, speedB, pos)};
		}
	}
	return tuple <double, double> {speedA, 0.0};
}

void Drive::run () {

	double position = 0.0;
	double speed = 0.0;
	tuple <double, double, double> section;
	tuple <double, double> look;

	double speedC, tmpdistance, delta_v, delta_distance;

	route.resetSection();
	while (route.isValidSection()) {
	
		section = route.getSection();
		
		double distance = get<0>(section); 
		double speedA = get<1>(section); 
		double speedB = get<2>(section); 

		if (speedA <= speedB) {
			motion(speed, speedA, distance, position);
			position += distance;
			speed = drive.back().speed;
		}
		else if (speedA > speedB) {
			look = getOutlookA(speed, speedB, distance, position);
			speedC = get<0>(look); 
			tmpdistance = get<1>(look); 

			motion(speed,speedC, tmpdistance, position);
			position += tmpdistance;
			speed = drive.back().speed;

			motion(speed,speedB,distance-tmpdistance, position);
			position += distance - tmpdistance;
			speed = drive.back().speed;
		}
		route.nextSection();
	}

	vector<DriveMark>::size_type i;
	for (i=0; i < (drive.size() - 1); ++i) {
		delta_v = (drive[i+1].speed - drive[i].speed) / 2.0 + drive[i].speed;
		delta_distance = drive[i+1].distance - drive[i].distance;
		drive[i+1].time = fabs(2 * delta_distance / delta_v) + drive[i].time;
	}

}

void Drive::writeResults () {

	FILE * file;
	vector<DriveMark>::size_type i;
		
	file = fopen("results/sample_A.drive", "w");
	for (i=0; i < drive.size(); ++i) {
		fprintf (file, "%e\t%e\t%e\n", drive[i].time, drive[i].distance/1000.0, drive[i]. speed);	
	}
	fclose (file);
}
